### ISSUE TITLE
Parametric types

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#192]
This have been discussed for probably about a year in various forms, so I've finally decided to make a distinct issue for it.

So, the proposal is "parametric types", i.e. instancing a types, passing additional parameters during instancing. Initial draft was in https://github.com/kaitai-io/kaitai_struct/issues/78#issuecomment-272581874, here is the definition:

```yaml
meta:
  id: bcd
params:
  - id: num_bytes
    type: u4
  - id: is_half_byte
    type: bool
seq:
  - id: digits
    type:
      switch-on: is_half_byte
      cases:
        true: b4
        false: u1
    repeat: expr
    repeat-expr: is_half_byte ? (num_bytes * 2) : num_bytes
```

And I propose invocation to be something like that:

```yaml
seq:
  - id: my_bcd
    type: bcd
    params:
      num_bytes: 4
      is_half_byte: true
```

and may be a brief positional form:

```yaml
seq:
  - id: my_bcd
    type: bcd(4, true)
```

## Possible use cases

* Pass constants to customize standard library types (as in this BCD example) and keep library from exponential growth of repetitive code
* Pass runtime-only values into a type (like `_index`), so it will know its position in some list
* Unification of TLV-based parsing for different files. For example, many network protocol parsers have `linktype` field, linktype enum and a switch like that:
```yaml
      switch-on: some_field
      cases:
        'linktype::ppi': packet_ppi
        'linktype::ethernet': ethernet_frame
        # ...
```
Right now we have to duplicate it (both a switch and enum) in every file. With parametric type, it can be moved into a separate .ksy with single parameter: that `some_field` that would be used to switch parsing.
* May be we could reuse this syntax for [other possible new custom type](https://github.com/kaitai-io/kaitai_struct/issues/188)

### Type definition

`type` in `params` definition is not the same `type` as in all other attributes. It actually only declares in-memory type, not in-stream type, so it doesn't have a concept of endianness, it doesn't have any terminators, it can't have fixed size, it can't have `process`, etc, etc. This difference is subtle, but important. We never actually had a type system like that before, so it has to be developed. My proposal is to keep it as close to existing in-stream type system as possible:

* `uX` for unsigned ints
* `sX` for signed ints
* `fX` for floats
  * `(le|be)` suffixes are not valid here
  * `enum: XXX` is allowed to specify enum-type-over-integer
* `str` for strings
  * `strz` is not valid
* no value = byte array
* `bool` for booleans
* `io` for KaitaiStream
* `struct` for arbitrary KaitaiStruct-based instance
* `any` for arbitrary any type (`Object`, `Any`, etc), if target language supports it

I'm not sure if it's a good idea to allow passing arrays, but probably we can think of some "generic" array form as well, if that would be needed. Like `repeat: true` or something.

### Definition and invocation

If there are any parameters defined, they would be normally prepended into any constructor(s), something like:

```cpp
bcd_t::bcd_t(uint32_t p_num_bytes, bool p_is_half_byte, kaitai::kstream *p__io, kaitai::kstruct *p__parent, bcd_t *p__root) : kaitai::kstruct(p__io) {
    m__parent = p__parent;
    m__root = this;

    // Store parameters
    m_num_bytes = p_num_bytes;
    m_is_half_byte = p_is_half_byte;
```

During invocation, a call like that is to be generated:

```cpp
m_foo = new bcd_t(4, true, m__io, this, m__root);
```

For now, I propose to implement all parameters as mandatory, mostly for initial simplicity. Later we can add something like `default: XXX` specification to allow default values (and thus optional parameters). Not passing a declared parameter or passing anything else beside declared parameters should be an error.

All opaque types are considered parameterless.

### Accessing

Parameters are to be treated exactly the same as eager seq attributes, the only difference is that they're received in constructor (and usually saved as fields of the object), not read in `_read` method. That means that relevant field declarations are to be generated, and relevant accessors. It should be possible to access parameters from other places in .ksy using normal `foo.bar.baz` syntax, i.e. something like `my_bcd.is_half_byte` should be legal.

Name clashes with other seq attributes / instance attributes are not allowed (check must be implemented).

**Comment by KOLANICH** [comment]
>uX for unsigned ints
>sX for signed ints

Do we need full inheritance? How we'll we deal with covariance/contravariance in this case (to deal with it we'll need some "template variables" to store types, but ks doesn't have this feature)? Or should we just have duck typing without parameters type checks on instantiation (with type checks on accessing properties/methods)?

**Comment by GreyCat** [comment]
@KOLANICH Sorry, I don't quite understand most of what you've wrote. What covariance/contravariance has to do with all this stuff, given that we don't have anything remotely resembling generics / templates, and especially more complex stuff like covariant/contravariant generics?

The system that I've proposed is basically what KS compiler uses right now internally (see [DataType.scala](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/datatype/DataType.scala)). KS expression language is already strongly typed, and I don't really think there's an easy way to change that (given that we have some target languages which do not even have reflection support or anything like that).

**Comment by KOLANICH** [comment]
>Sorry, I don't quite understand most of what you've wrote. What covariance/contravariance has to do with all this stuff, given that we don't have anything remotely resembling generics / templates, and especially more complex stuff like covariant/contravariant generics?

Unfortunately for now we don't have this. But
>instancing a types, passing additional parameters during instancing.

sounds like templates/generics like the ones we have in c++.

Types like
>uX for unsigned ints
sX for signed ints

sounds like abstract base classes and the types using them as parameters sounds like templates/generics with restrictions.

>struct for arbitrary KaitaiStruct-based instance
>any for arbitrary any type (Object, Any, etc), if target language supports it

sounds like base classes of fully object-oriented languages.

If you mean that these parameters are like functions in the sense arguments are taken from the stream, I don't understand why do we need that because it can be achieved with creating an additional type in most cases. In your example the parameters are fixed on compile time which means it is a template.

**Comment by GreyCat** [comment]
> sounds like templates/generics like the ones we have in c++.

Not really. Not we're just discussion passing of parameters into the constructor. Templates/generics is about allowing to passing a type into the very definition of the type.

> If you mean that these parameters are like functions in the sense arguments are taken from the stream in runtime, I don't understand why do we need that because it can be achieved with creating some additional types in most cases.

Please re-read "possible use cases" above. Yes, in *some* cases, we can get away by creating several pre-defined types. In many other cases, it's either impossible or at least very impractical.

> And I don't understand why do we need reflection, for example in C++ you can use templates and constexprs, including inherited ones, without reflection.

Anyway, generic/templates are definitely out of scope of this issue. If you want to discuss them - let us move to #135. You have a proposal of syntax there, but I have no idea how is it supposed to map to real-life languages.

**Comment by GreyCat** [comment]
I've started the implementation:

* [compiler, "param_types" branch](https://github.com/kaitai-io/kaitai_struct_compiler/tree/param_types)
* [tests, "param_types" branch](https://github.com/kaitai-io/kaitai_struct_tests/tree/param_types)

One new test was added so far: [param_def.ksy](https://github.com/kaitai-io/kaitai_struct_tests/blob/param_types/formats/params_def.ksy) — it tests definition of a type with parameters, instancing of such a class is done manually from a spec. C++, C#, Java, JS, Python, Ruby is implemented, PHP and Perl are probably coming next.

**Comment by GreyCat** [comment]
Ok, calling parametric types in implemented as well, only "short" form, i.e. like `type: foo(1, 2, "x")`.

Everything except for Perl works, and Perl is super strange, as I can't get the `param_def` test to fail originally (i.e. without any changes being applied to make it work).

**Comment by GreyCat** [comment]
Given that no one complained, I've merged parametric types to master. There is actually a huge problem with using `_index` now, as evidenced by new `IndexToParam*` family of tests, as it turned out that many language actually do *not* have an index variable in many types of loops.

**Comment by GreyCat** [comment]
I consider this done and it will be released as part of v0.8. All further improvements/changes would probably be opened as distinct tickets.
